// regexp-worker/lib/Procedures/procExecRegExp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResponseExecRegExp = exports.createRequestExecRegExp = exports.procExecRegExp = exports.isExecRegExpResponse = exports.isExecRegExpRequest = exports.requestTypeExecRegExp = void 0;
const util_1 = require("util");
const evaluateRegExp_1 = require("../helpers/evaluateRegExp");
const procedure_1 = require("./procedure");
exports.requestTypeExecRegExp = 'ExecRegExp';
exports.isExecRegExpRequest = (0, procedure_1.genIsRequest)(exports.requestTypeExecRegExp);
exports.isExecRegExpResponse = (0, procedure_1.genIsResponse)(exports.requestTypeExecRegExp);
function procExecRegExp(r) {
    if (!(0, exports.isExecRegExpRequest)(r))
        return undefined;
    try {
        const regex = (0, evaluateRegExp_1.toRegExp)(r.data.regexp);
        return createResponseExecRegExp(r, (0, evaluateRegExp_1.execRegExp)(regex, r.data.text));
    }
    catch (e) {
        return (0, procedure_1.createErrorResponse)(r, (0, util_1.format)(e));
    }
}
exports.procExecRegExp = procExecRegExp;
function createRequestExecRegExp(data) {
    return (0, procedure_1.createRequest)(exports.requestTypeExecRegExp, data);
}
exports.createRequestExecRegExp = createRequestExecRegExp;
function createResponseExecRegExp(request, data) {
    return (0, procedure_1.createResponse)(request.id, request.requestType, data);
}
exports.createResponseExecRegExp = createResponseExecRegExp;
