// regexp-worker/lib/Procedures/procedure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isErrorResponse = exports.createErrorResponse = exports.responseTypeError = exports.createResponse = exports.createRequest = exports.isResponse = exports.genIsResponse = exports.isRequest = exports.genIsRequest = void 0;
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
const uniqueId_1 = require("./uniqueId");
function genIsRequest(key) {
    return (v) => {
        return isRequest(v) && v.requestType === key;
    };
}
exports.genIsRequest = genIsRequest;
function isRequest(v) {
    return !!(typeof v === 'object' && typeof v.requestType === 'string' && (0, uniqueId_1.isId)(v.id));
}
exports.isRequest = isRequest;
function genIsResponse(key) {
    return (v) => {
        return isResponse(v) && v.responseType === key;
    };
}
exports.genIsResponse = genIsResponse;
function isResponse(v) {
    return !!(typeof v === 'object' && typeof v.responseType === 'string' && (0, uniqueId_1.isId)(v.id));
}
exports.isResponse = isResponse;
function createRequest(requestType, data) {
    return { id: (0, uniqueId_1.createId)(), requestType, data };
}
exports.createRequest = createRequest;
function createResponse(id, responseType, data) {
    return { id, timestamp: Date.now(), responseType, data };
}
exports.createResponse = createResponse;
exports.responseTypeError = 'Error';
function createErrorResponse(request, message, error) {
    if (!isRequest(request)) {
        return createResponse((request === null || request === void 0 ? void 0 : request.id) || uniqueId_1.NullID, exports.responseTypeError, {
            requestType: request === null || request === void 0 ? void 0 : request.requestType,
            message,
            error,
        });
    }
    const { id, requestType } = request;
    return createResponse(id, exports.responseTypeError, {
        requestType,
        message,
        error,
    });
}
exports.createErrorResponse = createErrorResponse;
exports.isErrorResponse = genIsResponse(exports.responseTypeError);
