// regexp-worker/lib/Procedures/procEcho.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResponseEcho = exports.createRequestEcho = exports.procEcho = exports.isEchoResponse = exports.isEchoRequest = exports.typeEcho = void 0;
const procedure_1 = require("./procedure");
exports.typeEcho = 'Echo';
exports.isEchoRequest = (0, procedure_1.genIsRequest)(exports.typeEcho);
exports.isEchoResponse = (0, procedure_1.genIsResponse)(exports.typeEcho);
function procEcho(r) {
    if (!(0, exports.isEchoRequest)(r))
        return undefined;
    if (typeof r.data !== 'string') {
        return (0, procedure_1.createErrorResponse)(r, 'Empty Echo');
    }
    return createResponseEcho(r, r.data);
}
exports.procEcho = procEcho;
function createRequestEcho(data) {
    return (0, procedure_1.createRequest)(exports.typeEcho, data);
}
exports.createRequestEcho = createRequestEcho;
function createResponseEcho(request, data) {
    return (0, procedure_1.createResponse)(request.id, request.requestType, data);
}
exports.createResponseEcho = createResponseEcho;
