// regexp-worker/lib/Procedures/procMatchRegExp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResponseMatchRegExp = exports.createRequestMatchRegExp = exports.procMatchRegExp = exports.isMatchRegExpResponse = exports.isMatchRegExpRequest = exports.requestTypeMatchRegExp = void 0;
const util_1 = require("util");
const evaluateRegExp_1 = require("../helpers/evaluateRegExp");
const procedure_1 = require("./procedure");
exports.requestTypeMatchRegExp = 'MatchRegExp';
exports.isMatchRegExpRequest = (0, procedure_1.genIsRequest)(exports.requestTypeMatchRegExp);
exports.isMatchRegExpResponse = (0, procedure_1.genIsResponse)(exports.requestTypeMatchRegExp);
function procMatchRegExp(r) {
    if (!(0, exports.isMatchRegExpRequest)(r))
        return undefined;
    try {
        const regex = (0, evaluateRegExp_1.toRegExp)(r.data.regexp);
        const regexResult = (0, evaluateRegExp_1.matchRegExp)(r.data.text, regex);
        return createResponseMatchRegExp(r, regexResult);
    }
    catch (e) {
        return (0, procedure_1.createErrorResponse)(r, (0, util_1.format)(e));
    }
}
exports.procMatchRegExp = procMatchRegExp;
function createRequestMatchRegExp(data) {
    return (0, procedure_1.createRequest)(exports.requestTypeMatchRegExp, data);
}
exports.createRequestMatchRegExp = createRequestMatchRegExp;
function createResponseMatchRegExp(request, data) {
    return (0, procedure_1.createResponse)(request.id, request.requestType, data);
}
exports.createResponseMatchRegExp = createResponseMatchRegExp;
