// regexp-worker/lib/Procedures/procMatchRegExpArray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResponseMatchRegExpArray = exports.createRequestMatchRegExpArray = exports.procMatchRegExpArray = exports.isMatchRegExpArrayResponse = exports.isMatchRegExpArrayRequest = exports.requestTypeMatchRegExpArray = void 0;
const util_1 = require("util");
const evaluateRegExp_1 = require("../helpers/evaluateRegExp");
const procedure_1 = require("./procedure");
exports.requestTypeMatchRegExpArray = 'MatchRegExpArray';
exports.isMatchRegExpArrayRequest = (0, procedure_1.genIsRequest)(exports.requestTypeMatchRegExpArray);
exports.isMatchRegExpArrayResponse = (0, procedure_1.genIsResponse)(exports.requestTypeMatchRegExpArray);
function procMatchRegExpArray(r) {
    if (!(0, exports.isMatchRegExpArrayRequest)(r))
        return undefined;
    try {
        const regex = r.data.regexps.map((r) => (0, evaluateRegExp_1.toRegExp)(r));
        const regexResult = (0, evaluateRegExp_1.matchRegExpArray)(r.data.text, regex);
        return createResponseMatchRegExpArray(r, regexResult);
    }
    catch (e) {
        return (0, procedure_1.createErrorResponse)(r, (0, util_1.format)(e));
    }
}
exports.procMatchRegExpArray = procMatchRegExpArray;
function createRequestMatchRegExpArray(data) {
    return (0, procedure_1.createRequest)(exports.requestTypeMatchRegExpArray, data);
}
exports.createRequestMatchRegExpArray = createRequestMatchRegExpArray;
function createResponseMatchRegExpArray(request, data) {
    return (0, procedure_1.createResponse)(request.id, request.requestType, data);
}
exports.createResponseMatchRegExpArray = createResponseMatchRegExpArray;
